### PR TITLE
Meetings API Support

### PR DESCRIPTION
**Pull request by SecondeJK**
This PR adds support for the Meetings API.

## Description
A new Meetings Client can be fetched using `$client->meetings()`

## Motivation and Context
Evolution of SDK support for Vonage products.

## How Has This Been Tested?
New full test suite added for this client

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/Meetings/Client.php [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+        $response->setHydrator($hydrator);
+
+        // Currently have to do this until we can extend off the Iterator to handle Meetings data structures
+        $roomPayload = [];
+
+        foreach ($response as $room) {
+            $roomPayload[] = $room;
+        }
+
+        return $roomPayload;
```
Comment (manchuck) [inline src/Meetings/Client.php:241]:
```php
return array_map(
    function($room) {
        return $room
    },
    $response
);
```

File: src/Meetings/ClientFactory.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
+<?php
+
+/**
+ * Vonage Client Library for PHP
+ *
+ * @copyright Copyright (c) 2016-2022 Vonage, Inc. (http://vonage.com)
+ * @license https://github.com/Vonage/vonage-php-sdk-core/blob/master/LICENSE.txt Apache License 2.0
+ */
```
Comment (manchuck) [inline src/Meetings/ClientFactory.php:8]: See comment about file header

File: test/Meetings/ClientTest.php [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+
+    public function testWillGetAvailableRooms(): void
+    {
+        $this->vonageClient->send(Argument::that(function (RequestInterface $request) {
+            $this->assertEquals('GET', $request->getMethod());
+
+            $uri = $request->getUri();
+            $uriString = $uri->__toString();
+            // @TODO Pagination isn't actually being handled in a HAL way
+            $this->assertEquals('https://api-eu.vonage.com/beta/meetings/rooms?page_index=1', $uriString);
```
Comment (manchuck) [inline test/Meetings/ClientTest.php:73]: I think you mean to have the query parameter be `page_size`. If not replace `page_index` with `start_id`. That is how it is in the API defines it

File: test/Meetings/ClientTest.php [hunk 4]
```diff
@@ -0,0 +1,558 @@
...
+    }
+
+    public function testWillGetRoomsAssociatedWithThemeUsingFilter(): void
+    {
+        $this->vonageClient->send(Argument::that(function (RequestInterface $request) {
+            $this->assertEquals('GET', $request->getMethod());
+
+            $uri = $request->getUri();
+            $uriString = $uri->__toString();
+            $this->assertEquals('https://api-eu.vonage.com/beta/meetings/themes/323867d7-8c4b-4dce-8c11-48f14425d888/rooms?start_id=245&end_id=765&page_index=1', $uriString);
```
Comment (manchuck) [inline test/Meetings/ClientTest.php:516]: `page_index` should be  `page_size` based on the context

File: test/Meetings/Fixtures/Responses/get-upload-urls-success.json [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+      "Content-Type": "image/png",
+      "key": "auto-expiring-temp/logos/white/ca63a155-d5f0-4131-9903-c59907e53df0",
+      "logoType": "abc123",
+      "bucket": "abc123",
+      "X-Amz-Algorithm": "abc123",
+      "X-Amz-Credential": "abc123",
+      "X-Amz-Date": "abc123",
+      "X-Amz-Security-Token": "abc123",
+      "Policy": "abc123",
+      "X-Amz-Signature": "abc123"
```
Comment (manchuck) [inline test/Meetings/Fixtures/Responses/get-upload-urls-success.json:14]: I would make these fields distinct to ensure that they are being passed correctly

File: test/Psr7AssertionTrait.php [hunk 6]
```diff
@@ -11,6 +11,8 @@
 
 namespace VonageTest;
 
+use _PHPStan_582a9cb8b\Symfony\Component\Finder\Iterator\RecursiveDirectoryIterator;
```
Comment (manchuck) [inline test/Psr7AssertionTrait.php:14]: WTH ios PHPStan doing here?

File: test/Verify/ExceptionErrorHandlerTest.php [hunk 7]
```diff
@@ -1,10 +1,12 @@
 <?php
+
 /**
  * Vonage Client Library for PHP
  *
  * @copyright Copyright (c) 2020 Vonage, Inc. (http://vonage.com)
  * @license https://github.com/Vonage/vonage-php-sdk-core/blob/master/LICENSE.txt Apache License 2.0
  */
+
```
Comment (manchuck) [inline test/Verify/ExceptionErrorHandlerTest.php:9]: See comment about file header here

**Review by manchuck — CHANGES_REQUESTED** [review]
I don't see any way to upload the image to S3.

File: src/Client/Exception/NotFound.php [hunk 8]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+/**
+ * Vonage Client Library for PHP
+ *
+ * @copyright Copyright (c) 2016-2022 Vonage, Inc. (http://vonage.com)
+ * @license https://github.com/Vonage/vonage-php-sdk-core/blob/master/LICENSE.txt Apache License 2.0
+ */
```
[the remark this answers is not in the document]
Comment (SecondeJK) [reply src/Client/Exception/NotFound.php:8]: I feel sorry for you when you have to re-review this and I've pulled them all out, but yeah we should bin it. There's no need for it, the license is included in the repo.

File: src/Client/Signature.php [hunk 9]
```diff
@@ -41,10 +41,12 @@ class Signature implements \Stringable
      * @throws ClientException
      */
     public function __construct(/**
-     * Params to Sign
-     */
-    protected array $params, $secret, $signatureMethod)
-    {
+         * Params to Sign
+         */
```
[the remark this answers is not in the document]
Comment (SecondeJK) [reply src/Client/Signature.php:38]: comment from 2016, useless

File: src/Meetings/Client.php [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+            $start_id ? $filterParams['start_id'] = $start_id : null;
+            $end_id ? $filterParams['end_id'] = $end_id : null;
+
+            if ($filterParams) {
+                $response = $this->api->search(new KeyValueFilter($filterParams), '/rooms');
+            }
+        } else {
+            $response = $this->api->search(null, '/rooms');
+        }
+
```
[the remark this answers is not in the document]
Comment (SecondeJK) [reply src/Meetings/Client.php:78]: We get pagination in, this becomes a lot nicer!

File: src/Meetings/Client.php [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+        $response->setHydrator($hydrator);
+
+        // Currently have to do this until we can extend off the Iterator to handle Meetings data structures
+        $roomPayload = [];
+
+        foreach ($response as $room) {
+            $roomPayload[] = $room;
+        }
+
+        return $roomPayload;
```
[the remark this answers is not in the document]
Comment (SecondeJK) [reply src/Meetings/Client.php:94]: Not sure I'd agree with memory management, but can make the code look cleaner

File: test/Meetings/ClientTest.php (hunk 3 above)
Comment (SecondeJK) [reply test/Meetings/ClientTest.php:75]: page_index is automatically added by the API Iteratorable class. For the time being, the backend will just ignore this query parameter and we might get defined pagination that will clean this up

File: test/Psr7AssertionTrait.php (hunk 6 above)
Comment (SecondeJK) [reply test/Psr7AssertionTrait.php]: bad merge

**Comment by SecondeJK** [comment]
@manchuck S3 now added, file headers cleaned up, not changed the `foreach` because `array_map` isn't used once in the codebase yet, and it would seem inconsistent here to just start using a different method of iteration when performance-wise there is no difference between the two from what I understand

**Comment by SecondeJK** [comment]
Will be replaced by new branch

**Pull request closed without merging**